### PR TITLE
support same padding in maxpooling and avgpooling

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

support same padding in maxpooling and avgpooling

## How was this patch tested?

add unit tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala [hunk 1]
```diff
@@ -29,6 +29,18 @@ import com.intel.analytics.bigdl.utils.Engine
...
+ *
+ * padAlongHeight = Math.max(0, (outHeight - 1) * strideH + kernelH - inHeight)
+ * padAlongWidth = Math.max(0, (outWidth - 1) * strideW + kernelW - inWidth)
+ *
+ * padTop = padAlongHeight / 2
+ * padLeft = padAlongWidth / 2
+ *
  * @param kW kernel width
  * @param kH kernel height
  * @param dW step width
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala]: update the java doc param list?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala [hunk 2]
```diff
@@ -186,6 +200,24 @@ class SpatialAveragePooling[@specialized(Float, Double) T: ClassTag](
     }
   }
 
+  // return (padTop, padDown, padLeft, padRight)
+  protected def getPadding(inputHeight: Int, inputWidth: Int): (Int, Int, Int, Int) = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala]: As discussed, move it to a util function

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala [hunk 3]
```diff
@@ -186,6 +200,24 @@ class SpatialAveragePooling[@specialized(Float, Double) T: ClassTag](
...
+    if (padW == -1 && padH == -1) {
+      // deal with SAME padding
+      val oW = Math.ceil(inputWidth.toFloat / dW.toFloat).toInt
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala]: is Math.ceil right when poolingLayer.ceilMode=false?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala (hunk 1 above)
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala]: It's just an explanation, no parameters added here

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala (hunk 3 above)
Comment (yangw1234) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialAveragePooling.scala]: It is correct actually. In the "SAME" padding algorithm, when padding is need, the algorithm will guarantee `inputHeight + padAlongH - kH` is always divisible by `dH` where the notion of `ceiling` and `floor` does not make sense. However, the code is a little confusing, I'll try to refactor that.

**Comment by yangw1234** [comment]
pr validation pass

**Comment by yiheng** [comment]
LGTM

**Pull request merged**